src/passes/Poppify.cpp

```diff
@@ -0,0 +1,504 @@
+        builder.makeLocalSet(elems[i], builder.makePop(types[i])));
+    }
+    if (curr->isTee()) {
+      // Use a tee followed by gets to retrieve the tuple
+      instrs.push_back(
+        builder.makeLocalTee(elems[0], builder.makePop(types[0]), types[0]));
+      for (size_t i = 1; i < types.size(); ++i) {
+        instrs.push_back(builder.makeLocalGet(elems[i], types[i]));
+      }
+    } else {
+      // Otherwise just add the last set
+      instrs.push_back(
+        builder.makeLocalSet(elems[0], builder.makePop(types[0])));
+    }
+  } else {
+    poppify(curr);
+    instrs.push_back(curr);
+  }
+}
+
+void Poppifier::emitGlobalGet(GlobalGet* curr) {
+  auto& instrs = scopeStack.back().instrs;
+  if (curr->type.isTuple()) {
+    auto types = module->getGlobal(curr->name)->type;
+    for (Index i = 0; i < types.size(); ++i) {
+      instrs.push_back(
+        builder.makeGlobalGet(getGlobalElem(curr->name, i), types[i]));
+    }
+  } else {
+    instrs.push_back(curr);
+  }
+}
+
+void Poppifier::emitGlobalSet(GlobalSet* curr) {
+  auto& instrs = scopeStack.back().instrs;
+  if (curr->value->type.isTuple()) {
+    auto types = module->getGlobal(curr->name)->type;
+    for (Index i = types.size(); i > 0; --i) {
+      instrs.push_back(builder.makeGlobalSet(getGlobalElem(curr->name, i - 1),
+                                             builder.makePop(types[i - 1])));
+    }
+  } else {
+    poppify(curr);
+    instrs.push_back(curr);
+  }
+}
+
+void Poppifier::poppify(Expression* expr) {
+  struct Poppifier : PostWalker<Poppifier> {
+    bool scanned = false;
+    Builder builder;
+    Poppifier(Builder& builder) : builder(builder) {}
+
+    static void scan(Poppifier* self, Expression** currp) {
+      if (!self->scanned) {
+        self->scanned = true;
+        PostWalker<Poppifier>::scan(self, currp);
+      } else {
+        *currp = self->builder.makePop((*currp)->type);
+      }
+    }
+  } poppifier(builder);
+  poppifier.walk(expr);
+}
+
+class PoppifyFunctionsPass : public Pass {
+  bool isFunctionParallel() override { return true; }
+  void
+  runOnFunction(PassRunner* runner, Module* module, Function* func) override {
+    if (func->profile != IRProfile::Poppy) {
+      Poppifier(func, module).write();
+      func->profile = IRProfile::Poppy;
+    }
+  }
+  Pass* create() override { return new PoppifyFunctionsPass; }
+};
+
+} // anonymous namespace
+
+class PoppifyPass : public Pass {
+  void run(PassRunner* runner, Module* module) {
+    PassRunner subRunner(runner);
+    subRunner.add(std::make_unique<PoppifyFunctionsPass>());
+    subRunner.add(std::make_unique<ReFinalize>());
+    subRunner.run();
+    lowerTupleGlobals(module);
+  }
+
+  void lowerTupleGlobals(Module* module) {
+    Builder builder(*module);
+    std::vector<std::unique_ptr<Global>> newGlobals;
+    for (int g = module->globals.size() - 1; g >= 0; --g) {
+      const auto& global = *module->globals[g];
+      if (global.type.isTuple()) {
```

Comment:
Nit: Maybe we can make this an early exit in case it is not a tuple, given that this loop has also other for and ifs below